### ISSUE TITLE
[16.0] How to Authenticate with oAuth?

### ISSUE THREAD

**Issue by luke-stdev001** [divad1196/odoo_graphql#9]
@divad1196 ,

Love this module, thanks for releasing it on the app store for free!

I'm trying to use this to build a Swift/Kotlin mobile app that authenticates with oAuth, and I have the Google and Facebook oAuth setup in Odoo, but can't figure out how I would authenticate with oAuth from GraphQL (ie. protect GraphQL endpoint with GraphQL and basic auth.

Are you available for some contract work in order to get this module migrated to 16.0 (it's already working in 16.0 btw), and help us to authenticate via oAuth directly from GraphQL, as well as teach us how to see GraphQL documentation (eg. use GraphiQL and see docs related to the endpoint)?

Please reach out to me at:
it@combuilt.com.au

to discuss.

**Comment by divad1196** [comment]
Hi @luke-stdev001 

Happy that the module is of some help.
I know, at least it was one of the goals when creating this module, that the module would migrate as it is. Thank you for confirming that.

I would be also happy to take your request into account to improve this module.

For the OAuth2 question, I first want to clarify the following point:
* The whole authentication is handled by Odoo. 
* OAuth2's goal is not to define a new authentication method but to have a way to authentify without giving your credentials directly to a server.
* Once OAuth2 is done, the process is the same as when authenticating with login/password. In the case of Odoo, you will have a cookie with the session_id.
* OAuth2, at least the worflow supported by Odoo, only need Odoo to provide 1 route for the redirection.

This means that, once your are connected (OAuth2 or standard connection) you end up in the exact same situation and the graphql api will work.

Can you clarify what you expect from graphql in the OAuth2 process?


For the GraphiQl tool, it was work in progress and there is currently a file downloadable but it needs to be integrated with the tool directly. I don't have a lot of time this week but I will carry on this development as soon as possible since someone is asking for it.

**Comment by luke-stdev001** [comment]
@divad1196 ,

Thanks for getting back to me. I will confirm all points with you as soon as possible.

**Comment by luke-stdev001** [comment]
@divad1196 ,

Would you be able to contact me via email at:

it@combuilt.com.au

when you can spare a minute? I'd like to discuss some possibilities with you regarding collaboration and funding towards this project.

**Comment by divad1196** [comment]
Hi @luke-stdev001 
I am not interested in being funded for what it implies.

Also, I prefer to have the discussions pn github so that anyone can have a look at what was discussed, and why, and can contribute to the discussion.

By the way, I made some advancement toward adding introspection to the project, I have been lacking time to finish it and will be pretty taken for the 2 next weeks. In 2-3 weeks, I should be able to take on and finish the development.

**Comment by luke-stdev001** [comment]
@divad1196 ,

What I was suggesting in regards to funding was purely to provide some financial backing for the project to help add certain features to this that we need (creation of a 16.0 branch, introspection, security hardening, and documentation for certain aspects such as authentication via oAuth2, etc.). If you do not want this, that's completely understandable, I just wanted to discuss the possibility with you if you were interested.

> By the way, I made some advancement toward adding introspection to the project, I have been lacking time to finish it and will be pretty taken for the 2 next weeks. In 2-3 weeks, I should be able to take on and finish the development.
This is fantastic news! Thanks for all the hard work on the module, it's very useful. I look forward to checking out the new features once they land in the 14.0 branch.

**Comment by divad1196** [comment]
@luke-stdev001 No offense taken, I totally understand why you proposed.

---
**TL;DR:** OAuth2 has nothing to do with this module.
For OAuth2, this is not the role of this module:
OAuth2 is not magic nor defines how the final server (Odoo in this case) manages authentication.

OAuth2 defines how a user proves he is who he is the first time he connects. Then, this is Odoo's responsibility to manage  the authentication. Odoo is both the resource server and the authorization server, meaning it doesn't need some token based reauthentication system but can rely on session based authentication.

---

The branch 16.0 works out of the box, I could have created it right away as you know, but I hoped to finish the introspection first.
I just created the branch 16.0 and published it on Odoo apps.

**Comment by divad1196** [comment]
@luke-stdev001 The feature is done and available on the store.
More information are provided in the module's `readme`  (e,.g Feature not active by default: performance and security).
This feature should only be used for development purpose.

Let me know if you see any issue that I missed

Nb: The same feature was requested in [issue #10](https://github.com/divad1196/odoo_graphql/issues/10),  I did the same answer to both of you, but I would prefer having only 1 github issue open.
Since this issue also mention fundings and OAuth2 and that those 2 topics are closed, I would prefer that any issue you find regarding the introspection is done on the other github issue (#10)
